### ISSUE TITLE
review and apply Jader H. Silva's patch for ngx.re.split()

### ISSUE THREAD

**Issue by chrislovecnm** [openresty/lua-nginx-module#217]
This is in the TODO's and would be very very helpful for a high throughput app that in is the works.  We are using gsub based splits which are SLOW ...

What is the timeframe?

Thanks

Chris

**Comment by jmealo** [comment]
@agentzh bump. This would be very helpful!

Kudos on openresty!

**Comment by bjoe2k4** [comment]
[njs](http://mailman.nginx.org/pipermail/nginx-devel/2016-July/008493.html) in the meantime got support for `String.split()` and is ahead of ngx_lua ;-)

**Comment by agentzh** [comment]
@bjoe2k4 Actually one can easily implement `ngx.re.split` with pure Lua with the existing `ngx.re` API.

**Comment by agentzh** [comment]
Nevertheless, we could provide a standard implementation in lua-resty-core via the new Lua module `ngx.re`. And it can be used like below:

```lua
local ngx_re = require "ngx.re"
local res, err = ngx_re.split(str, [[\s*,\s*]], "jo")
```

For this reason, I don't want to apply Jader H. Silva's patch for `ngx.re.split()`. Interested contributors can try creating a pull request for lua-resty-core to implement the API demonstrated above using the same FFI-based technique as in the `resty.core.regex` module.

**Comment by agentzh** [comment]
This way we don't have to maintain two implementations for `ngx.re.split` (one is based on the standard Lua 5.1's C API, one is based on LuaJIT's FFI API).

**Comment by agentzh** [comment]
For the best performance, we can also support an optional result table fed as an argument, as in

```lua
local ngx_re = require "ngx.re"
local my_table = {}
for i = 1, n do
     local res, err = ngx_re.split(str, regex, opts, my_table)
     if not res then ... end
     -- process the res table...
end
```

This way we do not have to create `n` result tables in the loop.

**Comment by agentzh** [comment]
Well, we can also support the optional `limit` argument as in perl's `split` builtin, as in

```lua
local res, err = ngx_re.split("a,b,c,d", ",", "jo", 2)
-- res is now {"a", "b,c,d"}
```

Another feature is that when the regex contains a sub-match capturing group, we return the sub-match capture in the separate pattern in the resulting Lua table as well, as in

```lua
local res, err = ngx_re.split("a,b,c,d", "(,)", "jo")
-- res is now {"a", ",", "b", ",", "c", ",", "d"}
```

**Comment by agentzh** [comment]
Patches welcome! Volunteers welcome!

**Comment by chrislovecnm** [comment]
Been open too long ... closing

**Comment by lifeblood** [comment]
ngx.re.split can be use now?

**Comment by thibaultcha** [comment]
I took a shot at this: https://github.com/openresty/lua-resty-core/pull/66

**Comment by agentzh** [comment]
I've merged the patch from @thibaultcha. The `split` function of the `ngx.re` module of `lua-resty-core` will get included in the next OpenResty (1.11.2.2) release (soon!).
